### PR TITLE
[SYCL] Enable host optimization of work-item free functions

### PR DESCRIPTION
**Pull request by cperkinsintel**
The SYCL free functions ( `this_item`, `this_id`, etc) are expensive to support on host devices. They cause performance delays because every iteration through one of the `parallel_for` routines the various indexing values have to be updated in case the users code might call `this_item` or `this_id` (or the others).   But with the new `callsThisItem`  method added to the Kernel Information (thanks @rdeodhar !), the host device can avoid paying the performance penalty if the users code doesn't actually call `this_item`..   We can detect at compile time whether or not any of the `this_xxx` free functions are used by the users code, and if not, don't bother storing the indexing data in each loop iteration. 

In this PR we add further expand the Kernel Information to support a `callsAnyThisFreeFunction` method, and we use it to avoid the sundry `store_item` etc. calls on the host.

### CODE REVIEW

File: sycl/include/CL/sycl/detail/cg_types.hpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include <CL/sycl/detail/host_profiling_info.hpp>
 #include <CL/sycl/detail/kernel_desc.hpp>
 #include <CL/sycl/group.hpp>
+//#include <CL/sycl/handler.hpp>
```
Comment (Fznamznon) [inline sycl/include/CL/sycl/detail/cg_types.hpp:14]:
```suggestion
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -3920,7 +3933,14 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
...
     O << "  static constexpr bool callsThisItem() { return ";
-    O << K.CallsThisItem << "; }\n";
+    O << K.FreeFunctionCalls.CallsThisItem << "; }\n";
+    O << "  __SYCL_DLL_LOCAL\n";
+    O << "  static constexpr bool callsAnyThisFreeFunction() { return ";
+    O << (K.FreeFunctionCalls.CallsThisId ||
+          K.FreeFunctionCalls.CallsThisItem ||
+          K.FreeFunctionCalls.CallsThisNDItem ||
+          K.FreeFunctionCalls.CallsThisGroup)
+      << "; }\n";
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:3943]: It feels that if the final information in the integration header doesn't say which exactly free function is called, we don't need four different flags and four identical handlers (I mean `SYCLIntegrationHeader::setCallsThisI` here) in front-end either.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -3979,10 +3999,28 @@ void SYCLIntegrationHeader::addSpecConstant(StringRef IDName, QualType IDType) {
...
 void SYCLIntegrationHeader::setCallsThisItem(bool B) {
   KernelDesc *K = getCurKernelDesc();
   assert(K && "no kernels");
-  K->CallsThisItem = B;
+  K->FreeFunctionCalls.CallsThisItem = B;
+}
+
+void SYCLIntegrationHeader::setCallsThisNDItem(bool B) {
+  KernelDesc *K = getCurKernelDesc();
+  assert(K && "no kernels");
```
Comment (keryell) [inline clang/lib/Sema/SemaSYCL.cpp:4016]:
```suggestion
  assert(K && "no kernel");
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -3979,10 +3999,28 @@ void SYCLIntegrationHeader::addSpecConstant(StringRef IDName, QualType IDType) {
...
+  K->FreeFunctionCalls.CallsThisNDItem = B;
+}
+
+void SYCLIntegrationHeader::setCallsThisGroup(bool B) {
+  KernelDesc *K = getCurKernelDesc();
+  assert(K && "no kernels");
```
Comment (keryell) [inline clang/lib/Sema/SemaSYCL.cpp:4022]:
```suggestion
  assert(K && "no kernel");
```

**Review by keryell — CHANGES_REQUESTED** [review]
You can also update the PR description which is completely cryptic.
Perhaps with "Enable host optimization of work-item free functions" for the title and a few paragraphs of what is happening here and why.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:3943]: @intel/llvm-reviewers-runtime , WDYT?

Comment (cperkinsintel) [reply clang/lib/Sema/SemaSYCL.cpp:3943]: Pre-existing code unrelated to this PR needs to know if `this_item` is called and that isn't concerned about the other free functions. The optimization in this PR just needs to know if any of them are called.  
My thoughts are that it's best to make a proper record of each free function now.  Recording only "this_item" and "anything" seems sloppy.  
On the API side, though, the functions match our current needs. The pre-existing `callsThisItem()` and `callsAnyThisFreeFunction()` are the two affordances, but can be easily expanded in the future if required.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -2721,11 +2721,24 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
       if (!Visited.insert(FD).second)
         continue; // We've already seen this Decl
 
-      // Check whether this call is to sycl::this_item().
+      // Check whether this call is to free functions (  sycl::this_item(),
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply clang/lib/Sema/SemaSYCL.cpp]: done

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (cperkinsintel) [reply clang/lib/Sema/SemaSYCL.cpp]: done.

File: sycl/include/CL/sycl/detail/cg_types.hpp (hunk 1 above)
Comment (cperkinsintel) [reply sycl/include/CL/sycl/detail/cg_types.hpp]: thanks.  done.

File: clang/include/clang/Sema/Sema.h [hunk 6]
```diff
@@ -386,7 +399,8 @@ class SYCLIntegrationHeader {
     SmallVector<KernelParamDesc, 8> Params;
 
     // Whether kernel calls this_item()
-    bool CallsThisItem;
+    // bool CallsThisItem;
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply clang/include/clang/Sema/Sema.h]: thanks.  Done.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:3943]: Do you expect it to be required?

**Review by elizabethandrews — COMMENTED** [review]
FE changes LGTM, other than the unresolved comment. I am alright with keeping the 4 different flags for now, if we expect to differentiate between free functions in the future.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (cperkinsintel) [reply clang/lib/Sema/SemaSYCL.cpp:3943]: I don't know of anything pending. But I'd wager at even odds on needing to know about `this_nd_item` or `this_group` usage in the future.

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:3943]: Alright thanks! I'm ok with the FE changes.

**Review by elizabethandrews — DISMISSED** [review]
FE changes LGTM.

File: sycl/include/CL/sycl/handler.hpp [hunk 7]
```diff
@@ -111,13 +107,6 @@ SuggestedArgType argument_helper(...);
...
 
-/// Helper struct to get a kernel name type based on given \c Name and \c Type
-/// types: if \c Name is undefined (is a \c auto_name) then \c Type becomes
-/// the \c Name.
-template <typename Name, typename Type> struct get_kernel_name_t {
-  using name = Name;
-};
-
 /// Specialization for the case when \c Name is undefined.
 template <typename Type> struct get_kernel_name_t<detail::auto_name, Type> {
```
Comment (v-klochkov) [inline sycl/include/CL/sycl/handler.hpp:111]: Why not to move this specialization to kernel header as it was done with the default/main version of get_kernel_name_t?

File: sycl/include/CL/sycl/detail/cg_types.hpp [hunk 8]
```diff
@@ -203,6 +203,12 @@ class HostKernel : public HostKernelBase {
   template <class ArgT = KernelArgType>
   typename detail::enable_if_t<std::is_same<ArgT, sycl::id<Dims>>::value>
   runOnHost(const NDRDescT &NDRDesc) {
+    using NameT =
+        typename detail::get_kernel_name_t<KernelName, KernelType>::name;
+    std::string KName = typeid(NameT *).name();
+    using KI = detail::KernelInfo<KernelName>;
+    constexpr bool StoreLocation = KI::callsAnyThisFreeFunction();
+
```
Comment (v-klochkov) [inline sycl/include/CL/sycl/detail/cg_types.hpp:211]: These 5 lines of code seem to be copy-pasted 4 times. Can it be moved into a function returning constexpr bool ?

Comment (cperkinsintel) [reply sycl/include/CL/sycl/detail/cg_types.hpp]: Good catch.
We won't need the NameT and the KName anymore. I deleted them, and now it's simply the `using KI` and the call to the constexpr function.

File: sycl/include/CL/sycl/handler.hpp (hunk 7 above)
Comment (cperkinsintel) [reply sycl/include/CL/sycl/handler.hpp]: good catch, thanks.  Done.

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM

**Review by v-klochkov — APPROVED** [review]
Looks Good To Me.
I also did not find any changes that would break backward compatibility.

**Review by rbegam — COMMENTED** [review]
sycl changes LGTM. 
please fix the pre-commit fails.

**Comment by pvchupin** [comment]
fail is flaky, @v-klochkov faced the same elsewhere and going to open issue.

**Pull request merged**